### PR TITLE
Robustness improvements

### PR DESCRIPTION
**Pull request by jglick**
- [X] [JENKINS-40613](https://issues.jenkins-ci.org/browse/JENKINS-40613): make sure `FilePath.isDirectory` calls are subject to timeout
- [X] [JENKINS-37730](https://issues.jenkins-ci.org/browse/JENKINS-37730): include more diagnostics in `getStatus`
- [X] [JENKINS-38769](https://issues.jenkins-ci.org/browse/JENKINS-38769): make sure `stop` does something
- [X] [JENKINS-37486](https://issues.jenkins-ci.org/browse/JENKINS-37486): reported `NullPointerException` perhaps due to Guice problems

@reviewbybees

### CODE REVIEW

**Comment by recampbell** [comment]
Thanks Jesse. What use-cases can be covered by tests here?

**Comment by stephenc** [comment]
Would be good if you had a test that verified upgrading a system while a durable task is running does not lead to a zombie durable task ;-)

:bee:

**Comment by jglick** [comment]
> a test that verified upgrading a system while a durable task is running does not lead to a zombie durable task

Well, it depends on the nature of the upgrade. The most dangerous changes would be to the serialized form of a `StepExecution` (transitively, so including nested objects such as a `Controller` in this case). This patch should only be touching `transient` fields, though.

**Comment by rsandell** [comment]
:bee:

File: src/main/java/org/jenkinsci/plugins/workflow/steps/durable_task/DurableTaskStep.java [hunk 1]
```diff
@@ -120,42 +125,50 @@ public FormValidation doCheckReturnStatus(@QueryParameter boolean returnStdout,
...
+
     }
 
     /**
      * Represents one task that is believed to still be running.
      */
-    @Restricted(NoExternalUse.class)
     @SuppressFBWarnings(value="SE_TRANSIENT_FIELD_NOT_RESTORED", justification="recurrencePeriod is set in onResume, not deserialization")
-    public static final class Execution extends AbstractStepExecutionImpl implements Runnable {
+    static final class Execution extends StepExecution implements Runnable {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/durable_task/DurableTaskStep.java]: Unfortunately, this does seem to cause a `InvalidClassException`:

    Class does not extend stream superclass

Looking to see if there is some way around that which does not require going back to `AbstractStepExecutionImpl`.

Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/steps/durable_task/DurableTaskStep.java]: Filed as [JENKINS-40909](https://issues.jenkins-ci.org/browse/JENKINS-40909).

**Pull request merged**